distsql/select_result.go

```diff
@@ -74,28 +74,30 @@ type selectResult struct {
 	copPlanIDs []fmt.Stringer
 	rootPlanID fmt.Stringer
 
-	fetchDuration time.Duration
-	memTracker    *memory.Tracker
+	fetchDuration    time.Duration
+	durationReported bool
+	memTracker       *memory.Tracker
 }
 
 func (r *selectResult) Fetch(ctx context.Context) {
 }
 
 func (r *selectResult) fetchResp(ctx context.Context) error {
-	startTime := time.Now()
 	defer func() {
-		duration := time.Since(startTime)
-		r.fetchDuration += duration
-		if r.selectResp == nil {
+		if r.selectResp == nil && !r.durationReported {
 			// final round of fetch
 			// TODO: Add a label to distinguish between success or failure.
 			// https://github.com/pingcap/tidb/issues/11397
 			metrics.DistSQLQueryHistgram.WithLabelValues(r.label, r.sqlType).Observe(r.fetchDuration.Seconds())
```

Comment:
Could you put these into line 104, since it is not quite necessary using `defer` here.

Reply:
NVM, it's ok with me.